### ISSUE TITLE
Support Postgrex v0.14.0

### ISSUE THREAD

**Issue by slashdotdash** [commanded/eventstore#141]
Attempting to compile EventStore with the latest version of Postgrex published to Hex (v0.14.0) fails with the following compile error:

```
Generated postgrex app
==> eventstore
Compiling 49 files (.ex)

== Compilation error in file lib/event_store/storage/reader.ex ==
** (CompileError) lib/event_store/storage/reader.ex:73: Postgrex.Timestamp.__struct__/0 is undefined, cannot expand struct Postgrex.Timestamp
    lib/event_store/storage/reader.ex:73: (module)
could not compile dependency :eventstore, "mix compile" failed. You can recompile this dependency with "mix deps.compile eventstore", update it with "mix deps.update eventstore" or clean it with "mix deps.clean eventstore"
```

This is because [`Postgrex.Timestamp` has been deprecated and removed](https://github.com/elixir-ecto/postgrex/blob/master/CHANGELOG.md#v0140-2018-10-29) from Postgrex which is used in `EventStore.Storage.Reader`. The usage must be made optional (e.g. `Code.ensure_loaded?(Postgrex.Timestamp)`) to support both v0.13.0 and older and the latest v0.14.0.

**Comment by slashdotdash** [comment]
Postgrex v0.14.0 expects the [JSON encoder to provide an `encode_to_iodata!/1` function](https://github.com/elixir-ecto/postgrex/issues/282). This function has been [removed from Poison in v4.x](https://github.com/devinus/poison/issues/172) which means that Poison cannot be directly used with the latest Postgrex.

This is needed for EventStore to support the `jsonb` data type (for event data & metadata).

**Comment by vasspilka** [comment]
@slashdotdash What would you consider as a valid change for this? Would maybe a migration to Jason + a version bump to `0.16` be accepted as a solution? Or maybe just drop support for Poison 4? Jason does not have the `as: ` but one can use `struct/2` to achieve the same result.

**Comment by slashdotdash** [comment]
As both Postgrex and Phoenix have migrated from Poison to Jason for JSON encoding then EventStore can do the same. In which case we should deprecate support for Poison and previous Postgrex versions (v0.13.0 and older).

This will also impact Commanded since it uses Poison for event data & metadata deserialization (in [`Commanded.Serialization.JsonSerializer`](https://github.com/commanded/commanded/blob/master/lib/commanded/serialization/json_serializer.ex)).

**Comment by vasspilka** [comment]
Yes, I am aware, I already have written a Jason serializer, so when I get the time I'll look into these issues and send a PR :blue_heart: :purple_heart: :heart:

**Comment by raarts** [comment]
Hi, experiencing this problem. What is the recommended workaround, if I can't wait for this PR to be merged and a new release been published? Thanks.

**Comment by slashdotdash** [comment]
@raarts Unfortunately there is no workaround since Postgrex v0.14.0 has broken the integration with Poison for JSON serialization. We need to migrate EventStore to use Jason.

**Comment by raarts** [comment]
None at all? Maybe I can go back to an older Postgrex version (I *am* using phoenix 1.4 though)

**Comment by slashdotdash** [comment]
Phoenix doesn't have a dependency on Postgrex. I'm assuming you are also using Ecto v3 which _does_ depend upon the latest Posgrex release.

**Comment by raarts** [comment]
@slashdotdash I tried but got stuck in compile errors/dependency problems. I'll have to wait. Any timeline?

**Comment by raarts** [comment]
Just in case someone else has the same problem, this worked:

```
mix archive.install https://github.com/phoenixframework/archives/raw/master/phx_new-1.3.4.ez
mix phx.new eventsourcing
mix archive.install hex phx_new 1.4.0
```

(Did not find a way to keep two different versions of that mix task on my machine and pick between them)

**Comment by raarts** [comment]
Well, I just tried again, but:

```
Failed to use "postgrex" because
  apps/api/mix.exs requires ~> 0.13.0
  ecto_sql (version 3.0.0) requires ~> 0.14.0
```

Adding override: true to the postgrex dependency resulted in:

```
Failed to use "db_connection" (versions 2.0.0 to 2.0.3) because
  ecto_sql (version 3.0.0) requires ~> 2.0
  postgrex (versions 0.13.0 to 0.13.5) requires ~> 1.1
```

Next I changed the mix.exs file in deps/ecto_sql to require 1.0 with override, but running `mix deps.get` after that change did not seem to pick that up, was still complaining about requiring 2.0. Tried editing mix.lock file, but also didn't get picked up, and then I was lost.

I remember an earlier attempt ending up in compile errors. So I guess it's not as easy as it looks.

**Comment by vasspilka** [comment]
Yes, yes, its Ecto 3 causing that not much to do with phoenix as slashdot
noted. We are working on a migration but seems no one really has the time
for it at the moment..

**Comment by vasspilka** [comment]
There is no other `ecto_sql` other than the v 3 one. Before that it was just `ecto`, with version 3 it got split into 2 dependencies. You can find more here http://blog.plataformatec.com.br/2018/10/a-sneak-peek-at-ecto-3-0-breaking-changes/.

**Comment by darwinphi** [comment]
What's the solution for this?

**Comment by slashdotdash** [comment]
The blocking issue is to migrate JSON serialization from Poison to Jason. This is a significant and breaking change which needs to be carefully implemented. Unfortunately I don’t have time right now to do so.

Can you continue using Postgrex v0.13.x for now?

**Comment by davich** [comment]
We're continuing to use Phoenix 1.3 and Postgrex 0.13 for now. Eagerly awaiting an update :) Thanks

**Comment by slashdotdash** [comment]
@sleepiejohn You need to use Postgrex v0.13.5

**Comment by lewisf** [comment]
Even though everyone has different setups, hoping anyone might find this helpful.

Here are the packages I had to downgrade (running multiple phoenix 1.4 apps in umbrella):

removed `ecto_sql 3.0`
downgraded `ecto 3.0` to `ecto 2.2`
downgraded `phoenix_ecto 4.0` to `phoenix_ecto 3.6`
downgraded `postgrex 0.14.0` to `postgrex 0.13.5`
downgraded `guardian_db from 2.0` to `guardian_db 1.1` (2.0 relied on ecto 3.0)

**Comment by Adzz** [comment]
Am I right in thinking this is done now? Looks like there are commits in master switching from poison to Jason

**Comment by slashdotdash** [comment]
Yes, it's done in #143 and included in the v0.16.0 release published to Hex.
